Documenting Hub online synchronization (#3921)
uyuni-project/uyuni-docs: modules/specialized-guides/pages/large-deployments/hub-online-sync.adoc

[[hub-online-sync]]
= Hub Online Synchronization

//OM 2025-04-28: shall we write hub or HUb, peripheral or Peripheral? In the GUI, I saw mixed cases.

== Introduction 

Hub online synchronization reuses the existing repository syncing mechanism and synchronizes channels in the peripheral servers from the repositories on the hub server.

> Once the connection between hub and peripheral servers is established, the hub server becomes the main source of data for the peripheral server, effectively replacing {scclongform}.

Reviewer: >the hub server becomes the main source of data for the peripheral server, effectively replacing {scclongform}.

Not sure if we need to clarify, but this is not limited to vendor channels retrieved from SCC, but it also applies to custom channels. Once they are synced, the peripheral will fetch the packages from the hub and not from the original location of the custom channel defined on the hub.

The main characteristics of this feature are:

* There can only be one hub server per connection, with one or more peripheral servers.

* Each peripheral server can only have one hub server. 

* Content can be synchronized on regular basis, or on demand.

== Registration of the hub and peripheral servers

Hub online synchronization is configured from menu menu:Admin[Hub Configuration].

Configuration process uses token which uniquely identifies peripheral server's connection to the hub.

There are two ways to register a peripheral server to the hub server:

. by using a combination of token creation on the peripheral, and subsequent registration on the hub server.
  This method uses <<peripheral-token-generation>> and <<token-transfer>>.

> . by direct registration from the hub, without any interactions with the peripheral server. 

Reviewer: ```suggestion
. by direct registration from the hub, without any user interactions with the peripheral server. 
```

Maybe it's better to clarify it's without **user** interaction or the reader might think this registration method can be done fully on the hub without needing the peripheral to be up and reachable. With this method the hub is basically using API calls to perform the first steps of the token generation procedure.

[[token-transfer]]
.Procedure: Registering to the hub server with the token
. On the hub server, go to menu:Hub Configuration[Peripherals Configuration].
. Click button btn:[Add peripheral].
  A new form [literal]``Register a new peripheral server`` opens.
. In the field [literal]``Peripheral Server FQDN`` enter the name of the peripheral server.
. In the field [literal]``Registration mode`` select option [literal]``Existing token``.
. In the field [literal]``Token`` paste the token that was created on the peripheral server.
. In the field [literal]``Root CA certificate`` specify the certificate using one of the options:
  * Use option [literal]``Not needed`` if both hub and peripheral servers have the same certificate authority.
  * Use option [literal]``Upload a file`` if the servers have different certificate authorities to upload a certificate file.
  * Use option [literal]``Paste a PEM certificate`` to paste a certificate.  
. Click button btn:[Register].
  A newly registered peripheral serevr will appear on screen [literal]``Peripherals Configuration``.

=== Registration from the hub server directly

> It is possible initiate the registration of a peripheral server from hub server, without any interaction with the peripheral server.

Author: ```suggestion
It is possible to initiate the registration of a peripheral server from hub server, without any interaction with the peripheral server.
```

[[direct-registration]]
.Procedure: Direct registering from the hub server

. On the hub server, go to menu:Hub Configuration[Peripherals Configuration].
. Click button btn:[Add peripheral].
  A new form [literal]``Register a new peripheral server`` opens.
. In the field [literal]``Peripheral Server FQDN`` enter the name of the peripheral server.
. In the field [literal]``Registration mode`` select option [literal]``Administrator User/Password``.

> . In the fields [literal]``Username`` and [literal]``Password`` enter the credentials for the peripheral server.

Reviewer 2: Please say, that these credentials must be from a "SUSE Manager Administrator" of the peripheral server.

. In the field [literal]``Root CA certificate`` specify the certificate using one of the options:
  * Use option [literal]``Not needed`` if both hub and peripheral servers have the same certificate authority.
  * Use option [literal]``Upload a file`` if the servers have different certificate authorities to upload a certificate file.
  * Use option [literal]``Paste a PEM certificate`` in cases when PEM certificate is used.
. Click button btn:[Register].

> . The newly reigistered peripheral server will be shown in the menu:Systems[System List] with the value [literal]``Foreign`` in the column [literal]``System Type``.

Reviewer: Small typo

```suggestion
. The newly registered peripheral server will be shown in the menu:Systems[System List] with the value [literal]``Foreign`` in the column [literal]``System Type``.
```

. To access its details, click on the peripheral server's name and select tab menu:Details[Peripheral Server].

Peripheral server uses hub to access the vendor channels and does not connect to the {scclongform} directly.
Therefore, if you open configured peripheral server's page menu:Admin[Setup Wizard > Organization Credentials], menu:Admin[Setup Wizard > Products] or menu:Admin[Setup Wizard > PAYG Connections], you will see a notification that this is peripheral server and its connections are managed via hub.

=== Access tokens

All existing tokens are shown in menu:Hub Configuration[Access Tokens].

A token can be [literal]``Consumed`` and [literal]``Issued``.

Consumed::
This type of token is generated on the peripheral server and received by the hub server to be used

Issued::

> This type of token is issued by the hub server to be used by the peripheral server.

Reviewer: This is written from the perspective of the hub. 

The description would be mirrored if looking at the same UI page on a peripheral. In that case, issued tokens are the one generated by the peripheral and sent to the hub, while consumed are those issued by the hub and used by the peripheral to establish a connection.

Author: Thank you @mackdk.

In addition to being deleted, a token can also be invalidated.

> //TO DO: What is the fundamental difference between these two operations? In both cases the connection between hub and periperal will be stopped. Can the invalidated token be re-validated?

Reviewer: Currently, the main difference is that deleting is only possible when the server associated with the token is not registered as hub or peripheral, and it cannot be undone. 

Deactivating could be a way to ensure no communications happen until a new token is generated (if the existing one is for some reason compromised) or the current one is reactivated.

=== Access hub server details from the peripheral server

Every peripheral server stores the information about its hub server.

[IMPORTANT]
====
A peripheral server can only have one hub server configured.
====

.Procedure: Accessing hub server details
. On the peripheral server, go to menu:Hub Configuration[Hub Details].
. On the screen [literal]``Hub Details`` find the information about the hub server.
.. Field [literal]``Server FQDN`` shows the hub server's FQDN.
.. Field [literal]``Registration date`` shows the time when the peripheral serevr was registsred to the hub server. 
.. Field [literal]``Last modified`` shows the time of the last saved configuration change. 
.. Field [literal]``Root Certificate Authority`` shows certificate details.
   To download, edit or delete the root certificate, clicking btn:[Download], btn:[Edit] or btn:[Delete] respectively.
   Deleting the certificate will break the connection between servers.
.. Field [literal]``GPG Public Key`` shows whether the GPG key has been configured for the hub server.
   For more information about GPG keys between hub and peripheral servers, see <<gpg-for-hub-online-sync>>.

> .. Field [literal]``Mirrir credentials`` is the username the peripheral server uses when connecting to the hub server to synchronize vendor channels.

Reviewer 2: typo:

```suggestion
.. Field [literal]``Mirror credentials`` is the username the peripheral server uses when connecting to the hub server to synchronize vendor channels.
```

   This username is generated automatically on the hub server, and then transmitted to the peripheral server during the registration phase.

[[gpg-for-hub-online-sync]]
==== GPG key usage with hub online synhronization

> When the packages are signed with GPG key, the key is automatically transmitted from hub to peripheral server.

Reviewer 2: ```suggestion
When the metadata on the Hub are signed with a GPG key, the public key is automatically transmitted from hub to peripheral server.
```

Author: Changed.

By default, {productname} is not signing meta data.

> Therefore, when the pripheral server is downloading data from the hub server, unless the customer has created their own GPG key then they have no way of checking if the downloaded rpms have valid signature.

Reviewer 2: ```suggestion
Therefore, when the peripheral server is downloading data from the hub server there is no way of checking if the downloaded metadata have a valid signature, unless the customer has created their own GPG key.
```

To enable checking of the data integrity, the GPG key needs to be created on the hub side.
When the peripheral server is configured to communicate with the hub, the public GPG key will then automatically be transferred to it.

Once the GPG key is created on the hub, field [literal]``GPG Public Key`` will be set to show that this serevr is using the GPG key.

> For more information about setting up own GPG key, see xref:administration:repo-metadata.adoc[].

Reviewer 2: I think you need to update the peripheral registration when you create the GPG key after you registered the peripheral.
I need to check how this can be done. Or Thomas knows it

   To download, edit or delete the root certificate, clicking btn:[Download], btn:[Edit] or btn:[Delete] respectively.
   Deleting the certificate will break the connection between servers.
.. Field [literal]``GPG Public Key`` shows whether the GPG key has been configured for the hub server.
   For more information about GPG keys between hub and peripheral servers, see <<gpg-for-hub-online-sync>>.
.. Field [literal]``Mirror credentials`` is the username the peripheral server uses when connecting to the hub server to synchronize vendor channels.
   This username is generated automatically on the hub server, and then transmitted to the peripheral server during the registration phase.

[[gpg-for-hub-online-sync]]
==== GPG key usage with hub online synhronization

When the metadata on the hub server are signed with a GPG key, the public key is automatically transmitted from hub to peripheral server.

By default, {productname} is not signing meta data.
Therefore, when the peripheral server is downloading data from the hub server there is no way of checking if the downloaded metadata have a valid signature, unless the customer has created their own GPG key.

> To enable checking of the data integrity, the GPG key needs to be created on the hub side.

Reviewer 3: ```suggestion
To enable checking of the data integrity, the GPG key needs to be created on the hub.
```

When the peripheral server is configured to communicate with the hub, the public GPG key will then automatically be transferred to it.

> Once the GPG key is created on the hub, field [literal]``GPG Public Key`` will be set to show that this serevr is using the GPG key.

Reviewer 3: It is also worth checking the titles about capitalization again.

For more information about setting up own GPG key, see xref:administration:repo-metadata.adoc[].

=== Deregister peripheral server

Deregistration can happen from both sides, from the hub or from the peripheral server.

.Procedure: Deregistering from the peripheral server
. Go to menu:Hub Configuration[Hub Details]. 
. Click btn:[Deregister].
. Confirm the operation by clicking btn:[Deregister] on the pop-up window.
. Page menu:Hub Configuration[Hub Details] is now empty.

.Procedure: Deregistering from the hub server
. Go to menu:Hub Configuration[Peripheral Configuration].
. Find the perpheral server on the list.
. Click btn:[Deregister] next to the peripheral server's name.
. The peripheral server is no longer shown on the list.

== Synchronize channels from hub to peripheral server

Synchronizing vendor channels for the configured hub and server is done via dedicated user interface.

.Procedure: Synchronizing channels from hub to peripheral server
. Go to menu:Admin[Hub Configuration > Peripherals Configuration].

> . In the field [literal]``Synchronized channels`` click on btn:[Edit channels].

Reviewer 3: I do not know where this button actually is. `field` sounds strange.  Maybe, "pane", "dialog", or "column"?  Or just:
```suggestion
. In [literal]``Synchronized channels`` click btn:[Edit channels].
```
